File: cpp/백준/집합과 맵/Union_find/1717 집합의 표현.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> parent;

int find_parent(int x) {
    if (parent[x] != x) {
        parent[x] = find_parent(parent[x]);
    }
    return parent[x];
}

void union_parent(int a, int b) {
    a = find_parent(a);
    b = find_parent(b);
    if (a < b) parent[b] = a;
    else parent[a] = b;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    parent.assign(n + 1, 0);
    for (int i = 0; i <= n; ++i) parent[i] = i;

    for (int i = 0; i < m; ++i) {
        int oper, a, b;
        cin >> oper >> a >> b;
        if (oper == 0) {
            union_parent(a, b);
        } else {
            if (find_parent(a) == find_parent(b)) cout << "YES" << "\n";
            else cout << "NO" << "\n";
        }
    }

    return 0;
}

File: python/백준/집합과 맵/Union_find/1717 집합의 표현.py
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline

n, m = map(int, input().split())

parent = [0] * (n + 1)

for i in range(n + 1):
    parent[i] = i

def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]
        
    
for _ in range(m):
    oper, a, b = map(int, input().split())
    
    if oper == 0:
        union_parent(parent, a, b)
    else:
        if find_parent(parent, a) == find_parent(parent, b):
            print("YES")
        else:
            print("NO")
    
